framework/src/samplers/Sampler.C

```diff
@@ -12,103 +12,79 @@
-    _var_names(getParam<std::vector<std::string>>("perturb_parameters")),
-    _current_sample(0),
-    _failed_runs(true)
+    _distribution_names(getParam<std::vector<DistributionName>>("distributions")),
+    _seed(getParam<unsigned int>("seed"))
 {
-  _var_dist_map.clear();
-  _probability_weight.clear();
-  _var_value_hist.clear();
-  if (_var_names.size() != _dist_names.size())
-    mooseError("The size of perturb_parameters (",
-               _var_names.size(),
-               ") != the size of distributions (",
-               _dist_names.size(),
-               ").");
-  if (!_var_names.empty())
-  {
-    for (unsigned int i = 0; i < _var_names.size(); ++i)
-    {
-      _var_dist_map[_var_names[i]] = &getDistributionByName(_dist_names[i]);
-      _var_value_map[_var_names[i]] = 0.0;
-      _var_value_hist[_var_names[i]] = std::vector<Real>();
-    }
-  }
+  _generator.seed(0, _seed);
+  for (const DistributionName & name : _distribution_names)
+    _distributions.push_back(&getDistributionByName(name));
+  setNumberOfRequiedRandomSeeds(1);
 }
 
 void
-Sampler::generateSamples()
+Sampler::execute()
 {
+  // Get the samples then save the state so that subsequent calls to getSamples returns the same
+  // random numbers until this execute command is called again.
+  getSamples();
+  _generator.saveState();
 }
 
-std::vector<std::string>
-Sampler::getSampledVariableNames()
+std::vector<DenseMatrix<Real>>
+Sampler::getSamples()
 {
-  return _var_names;
-}
-
-std::vector<Real>
-Sampler::getSampledValues(const std::vector<std::string> & variableNames)
-{
-  std::vector<Real> vals;
-  for (auto & var_name : variableNames)
-  {
-    Real val = getSampledValue(var_name);
-    vals.push_back(val);
-  }
-  return vals;
+  _generator.restoreState();
+  sampleSetUp();
+  std::vector<DenseMatrix<Real>> output(_distributions.size());
+  for (std::size_t i = 0; i < _distributions.size(); ++i)
+    output[i] = sampleDistribution(*_distributions[i], i);
+  sampleTearDown();
+  return output;
 }
 
 Real
-Sampler::getSampledValue(const std::string & variableName)
-{
-  std::map<std::string, Real>::iterator it;
-  it = _var_value_map.find(variableName);
-  if (it == _var_value_map.end())
-    mooseError(
-        "Could not find the parameter: ", variableName, " in the list of perturbed parameters!");
-  return it->second;
-}
-
-std::vector<Real>
-Sampler::getProbabilityWeights()
+Sampler::rand(const unsigned int index)
 {
-  return _probability_weight;
+  mooseAssert(index < _seeds.size(), "The seed number index does not exists.");
+  return _generator.rand(_seeds[index]);
 }
 
-bool
-Sampler::checkRuns()
+void
+Sampler::setNumberOfRequiedRandomSeeds(const std::size_t & number)
 {
-  return _failed_runs;
+  if (number == 0)
+    mooseError("The number of seeds must be larger than zero.");
+  _seeds.resize(number);
+  _seeds[0] = _seed;
+  _generator.seed(0, _seed);
+  for (std::size_t i = 1; i < number; ++i)
+  {
+    _seeds[i] = _seeds[0] + 1;
```

Comment:
This doesn't look right: All of the seeds but the first are the same and a fixed offset from the first seed?
Maybe you are generating seeds elsewhere? I'd use the `std::fill` method here if this is indeed correct. However, I believe this is the place where you need to seed the master generator, then generate seeds for all the other generators based on a single master seed.